Storage/src/Connection/Rest.php

```diff
@@ -600,4 +636,201 @@ protected function supportsBuiltinCrc32c()
     {
         return Builtin::supports(CRC32::CASTAGNOLI);
     }
+
+    /**
+     * Add the required retry function and send the request.
+     *
+     * @param string $resource resource name, eg: buckets.
+     * @param string $method method name, eg: get
+     * @param array $args
+     */
+    private function sendWithRetry($resource, $method, array $args)
+    {
+        $retryMap = [
+            'projects.resources.serviceAccount' => 'serviceaccount',
+            'projects.resources.hmacKeys' => 'hmacKey',
+            'bucketAccessControls' => 'bucket_acl',
+            'defaultObjectAccessControls' => 'default_object_acl',
+            'objectAccessControls' => 'object_acl'
+        ];
+        $retryResource = isset($retryMap[$resource]) ? $retryMap[$resource] : $resource;
+        $args['restRetryFunction'] = $this->getRestRetryFunction(
+            $retryResource,
+            $method,
+            $args,
+            $this->restRetryFunction
+        );
+
+        $args = $this->addRetryHeaderCallbacks($args);
+
+        return $this->send($resource, $method, $args);
+    }
+
+    /**
+     * Adds the callback methods to $args which amends retry hash and attempt
+     * count to the headers.
+     * @param array $args
+     *
+     * @return array
+     */
+    private function addRetryHeaderCallbacks(array $args)
+    {
+        $requestHash = Uuid::uuid4()->toString();
+        $args['restOnRetryExceptionFunction'] = function (
+            \Exception $e,
+            $currentAttempt,
+            &$arguments
+        ) use ($requestHash) {
+            // Since we the the last attempt number here, so incrementing it
+            // to get the current attempt count.
+            // We're adding a '2' and not '1' as we need to incorporate the initial
```

Comment:
It's required for us to number the request in an one based indexing, thus:
* Initial request => 1
* Retry 1 => 2
* Retry 2 => 3
* Retry 3 => 4

The retry attempt is incremented after all the exception handling is done and since this logic works as a part of exception handling, it receives the `$retryCount`(which is zero based) of the current on going request instead of the next attempt for which we are modifying the header _(that too not exactly modifying the Request, we just enqueue the modification in the`$args` array which will get applied is the request reached the guzzle layer., thus no immediate updation which seemed your worry for corner case when 3 retries happen.)_

Reply:
This is confusing how this is currently implemented. I have two suggestions that would help:

1. Rename `$currentAttempt` to `$retryAttempt` to match the variable name that's being passed in from `ExponentialBackoff`. This helps because `$retryAttempt` is always one less than `$currentAttempt`, so the logic would make more sense.
2. In `ExponentialBackoff`, call the modify header function as the last thing to happen instead of the first thing. This would mean `$retryAttempt++` will have been called already, and so there would be no need to increment the number again.